### PR TITLE
OTT-1160: Introduce Prometheus Metric Engine for PubMatic openwrap module

### PR DESCRIPTION
**Pull request by ashishshinde-pubm**

### CODE REVIEW

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+	platformLabel  = "platform"
+	endpointLabel  = "endpoint" // TODO- apiTypeLabel ?
+	apiTypeLabel   = "api_type"
+	impTypeLabel   = "imp_type" //TODO -confirm ?
+	adFormatLabel  = "adformat"
+	contentLabel   = "content" //TODO -confirm ?
+	nbrLabel       = "nbr"
+	errorLabel     = "error"
+	hostLabel      = "host" // combination of node:pod
+	methodLabel    = "method"
```
Comment (pm-viral-vala) [inline modules/pubmatic/openwrap/metrics/prometheus/prometheus.go:73]: can we keep label names consistent with wrapper tracker publisher id, profile id, etc?

**Review by pm-viral-vala — CHANGES_REQUESTED** [review]

File: metrics/config/metrics.go [hunk 2]
```diff
@@ -7,13 +7,14 @@ import (
...
 	"github.com/prebid/prebid-server/openrtb_ext"
+	"github.com/prometheus/client_golang/prometheus"
 	gometrics "github.com/rcrowley/go-metrics"
 	influxdb "github.com/vrischmann/go-metrics-influxdb"
 )
 
 // NewMetricsEngine reads the configuration and returns the appropriate metrics engine
 // for this instance.
-func NewMetricsEngine(cfg *config.Configuration, adapterList []openrtb_ext.BidderName, syncerKeys []string, moduleStageNames map[string][]string) *DetailedMetricsEngine {
+func NewMetricsEngine(cfg *config.Configuration, adapterList []openrtb_ext.BidderName, syncerKeys []string, moduleStageNames map[string][]string, promRegistry *prometheus.Registry) *DetailedMetricsEngine {
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/config/metrics.go]: yes, modified the code to pass generic MetricsRegistry

[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/config/metrics.go]: done

File: metrics/prometheus/prometheus.go [hunk 3]
```diff
@@ -212,15 +212,15 @@ const (
 )
 
 // NewMetrics initializes a new Prometheus metrics instance with preloaded label values.
-func NewMetrics(cfg config.PrometheusMetrics, disabledMetrics config.DisabledMetrics, syncerKeys []string, moduleStageNames map[string][]string) *Metrics {
+func NewMetrics(cfg config.PrometheusMetrics, disabledMetrics config.DisabledMetrics, syncerKeys []string, moduleStageNames map[string][]string, promRegistry *prometheus.Registry) *Metrics {
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/prometheus/prometheus.go]: done

[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/prometheus/prometheus.go]: done

File: metrics/prometheus/prometheus.go [hunk 4]
```diff
@@ -212,15 +212,15 @@ const (
...
 	standardTimeBuckets := []float64{0.05, 0.1, 0.15, 0.20, 0.25, 0.3, 0.4, 0.5, 0.75, 1}
 	cacheWriteTimeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
 	priceBuckets := []float64{250, 500, 750, 1000, 1500, 2000, 2500, 3000, 3500, 4000}
 	queuedRequestTimeBuckets := []float64{0, 1, 5, 30, 60, 120, 180, 240, 300}
 	overheadTimeBuckets := []float64{0.00005, 0.0001, 0.0005, 0.001, 0.002, 0.005, 0.01, 0.05}
 
 	metrics := Metrics{}
-	reg := prometheus.NewRegistry()
+	reg := promRegistry
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply metrics/prometheus/prometheus.go]: done

File: modules/moduledeps/deps.go [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
+	"github.com/prometheus/client_golang/prometheus"
+)
 
 // ModuleDeps provides dependencies that custom modules may need for hooks execution.
 // Additional dependencies can be added here if modules need something more.
 type ModuleDeps struct {
-	HTTPClient *http.Client
+	HTTPClient         *http.Client
+	PrometheusCfg      *config.PrometheusMetrics
+	PrometheusRegistry *prometheus.Registry
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply modules/moduledeps/deps.go]: introduced MetricRegistry and MetricsCfg which contains generic metric-engines data

File: modules/moduledeps/deps.go [hunk 6]
```diff
@@ -1,9 +1,16 @@
...
+	"github.com/prebid/prebid-server/config"
+	"github.com/prometheus/client_golang/prometheus"
+)
 
 // ModuleDeps provides dependencies that custom modules may need for hooks execution.
 // Additional dependencies can be added here if modules need something more.
 type ModuleDeps struct {
-	HTTPClient *http.Client
+	HTTPClient         *http.Client
+	PrometheusCfg      *config.PrometheusMetrics
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply modules/moduledeps/deps.go]: done

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 7]
```diff
@@ -257,7 +253,7 @@ func (m OpenWrap) handleBeforeValidationHook(
 			if err != nil || len(bidderParams) == 0 {
 				result.Errors = append(result.Errors, fmt.Sprintf("no bidder params found for imp:%s partner: %s", imp.ID, prebidBidderCode))
 				nonMapped[bidderCode] = struct{}{}
-				m.metricEngine.RecordSlotNotMappedErrorStats(rCtx.PubIDStr, bidderCode)
+				m.metricEngine.RecordPartnerConfigErrors(rCtx.PubIDStr, bidderCode, models.PartnerErrSlotNotMapped)
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply modules/pubmatic/openwrap/beforevalidationhook.go]: Done. Added profile label for this metric.

File: modules/pubmatic/openwrap/bidderparams/pubmatic.go [hunk 8]
```diff
@@ -89,7 +89,7 @@ func PreparePubMaticParamsV25(rctx models.RequestCtx, cache cache.Cache, bidRequ
 		if len(slots) != 0 { // reuse this field for wt and wl in combination with isRegex
 			matchedPattern = slots[0]
 		} else {
-			rctx.MetricsEngine.RecordMisConfigurationErrorStats(rctx.PubIDStr, string(openrtb_ext.BidderPubmatic))
+			rctx.MetricsEngine.RecordPartnerConfigErrors(rctx.PubIDStr, string(openrtb_ext.BidderPubmatic), models.PartnerErrMisConfig)
```
[the remark this answers is not in the document]
Comment (ashishshinde-pubm) [reply modules/pubmatic/openwrap/bidderparams/pubmatic.go]: Added profile label for this metric. 
But need to discuss this, there is difference between HB behaviour and this code. Not able to map this error to exact location. We have created OTT-1182 ticket to address such issues.

File: modules/pubmatic/openwrap/metrics/prometheus/prometheus.go (hunk 1 above)
Comment (ashishshinde-pubm) [reply modules/pubmatic/openwrap/metrics/prometheus/prometheus.go]: Modified few parameters to make them consistent with go-tracker labels

**Pull request closed without merging**